self_test/generic/complex_matrices_test/cr_complex_matrix_test.cc

```diff
@@ -0,0 +1,209 @@
+
+// Oomph-lib includes
+#include "generic.h"
+
+using namespace std;
+using namespace oomph;
+
+void print_complex_vector(Vector<complex<double>>& x)
+{
+  for (unsigned i = 0; i < x.size(); i++)
+  {
+    cout << x[i] << endl;
+  }
+}
+
+void print_cr_complex_matrix(CRComplexMatrix& M)
+{
+  for (unsigned i = 0; i < M.nrow(); i++)
+  {
+    for (unsigned j = 0; j < M.ncol(); j++)
+    {
+      cout << M(i, j) << ", ";
+    }
+    cout << endl;
+  }
+}
+
+int main()
+{
+  unsigned long n = 2;
+  unsigned long m = 4;
+
+  Vector<complex<double>> values(4);
+  values[0] = complex<double>(1, 1);
+  values[1] = complex<double>(2, 1);
+  values[2] = complex<double>(3, 1);
+  values[3] = complex<double>(4, 1);
+
+  Vector<int> column_index(4);
+  column_index[0] = 0;
+  column_index[1] = 2;
+  column_index[2] = 1;
+  column_index[3] = 3;
+
+  Vector<int> row_start(3); // length = n+1
+  row_start[0] = 0;
+  row_start[1] = 2;
+  row_start[2] = 4; // Fictitous index = NNz
+
+  Vector<complex<double>> values_square(5);
+  values_square[0] = complex<double>(1, 1);
+  values_square[1] = complex<double>(2, 1);
+  values_square[2] = complex<double>(3, 1);
+  values_square[3] = complex<double>(4, 1);
+  values_square[4] = complex<double>(5, 1);
+
+  Vector<int> column_index_square(5);
+  column_index_square[0] = 0;
+  column_index_square[1] = 2;
+  column_index_square[2] = 1;
+  column_index_square[3] = 0;
+  column_index_square[4] = 3;
+
+  Vector<int> row_start_square(5); // length = n+1
+  row_start_square[0] = 0;
+  row_start_square[1] = 2;
+  row_start_square[2] = 3;
+  row_start_square[3] = 4;
+  row_start_square[4] = 5; // Fictitous index = NNz
+
+  Vector<std::complex<double>> x(4);
+  x[0] = complex<double>(2, 2);
+  x[1] = complex<double>(-2, 3);
+  x[2] = complex<double>(0.5, 2.2);
+  x[3] = complex<double>(-1.4, 3);
+
+  Vector<std::complex<double>> rhs(4);
+  rhs[0] = complex<double>(1, 0);
+  rhs[1] = complex<double>(0, 1);
+  rhs[2] = complex<double>(2, 1);
+  rhs[3] = complex<double>(-3, 1);
+
+  Vector<std::complex<double>> soln(4);
+
+  Vector<double> value_double(5);
+  value_double[0] = double(1);
+  value_double[1] = double(2);
+  value_double[2] = double(3);
+  value_double[3] = double(4);
+  value_double[4] = double(5);
+
+  unsigned* method_ptr = new unsigned;
+
+  // test default constructor
+  CRComplexMatrix matrix_default();
+
+  // test full matrix constructor
+  CRComplexMatrix matrix(values, column_index, row_start, n, m);
+  cout << matrix.nrow() << endl;
+  cout << matrix.ncol() << endl;
+  print_cr_complex_matrix(matrix);
+
+  CRComplexMatrix matrix_square(
+    values_square, column_index_square, row_start_square, 4, 4);
+  cout << matrix_square.nrow() << endl;
+  cout << matrix_square.ncol() << endl;
+  print_cr_complex_matrix(matrix_square);
+
+  CRComplexMatrix matrix_result(
+    values_square, column_index_square, row_start_square, 4, 4);
+  print_cr_complex_matrix(matrix_result);
```

Comment:
I'm curious, is there any reason why the constructor on line 123 is passed `n` and `m` for the matrix size arguments but these one aren't? Saw this in other places too.

Reply:
No particular reason. These can't be the same as n and m as they need to be 4 for the hard coded matrix initialization to work. I know we don't like magic(is that right?) numbers in principle, but for a short, self contained testing script I don't hold myself to the same standards.

Reply:
> ...magic(is that right?)...

Yep!

> but for a short, self contained testing script I don't hold myself to the same standards.

😮 !!!

I think if code is to be shared with the `oomph-lib` community, it should all be held to the same high standards! (Mostly because I remember what it was like when I first starting reading/working on `oomph-lib` code; magic numbers obfuscate the code and are likely to cause confusion.) Please could you replace the values using your pre-defined variable or, if necessary, new variables? Thanks!

Reply:
Puneet is turning Germanic. But he's right, and remember that self-tests always double as demo drivers -- it's where I (and I'm sure others too) look for examples of how stuff is used in anger.

Reply:
> Puneet is turning Germanic

I've been indoctrinated by the best!